types.go

```diff
@@ -0,0 +1,160 @@
+type RootScope interface {
+	Scope
+
+	// Close Ceases periodic reporting of the root scope and subscopes
+	Close()
+}
+
+// Capabilities is a description of metrics reporting capabilities
+type Capabilities interface {
+	// Reporting returns whether the reporter has the ability to actively report
+	Reporting() bool
+
+	// Tagging returns whether the reporter has the capability for tagged metrics
+	Tagging() bool
+}
+
+// Snapshot is a snapshot of values since last report execution
+type Snapshot interface {
+	// Counters returns a snapshot of all counters since last report execution
+	Counters() map[string]CounterSnapshot
+
+	// Gauges returns a snapshot of all counters since last report execution
+	Gauges() map[string]GaugeSnapshot
+
+	// Timers returns a snapshot of all counters since last report execution
+	Timers() map[string]TimerSnapshot
+}
+
+// CounterSnapshot is a snapshot of a counter
+type CounterSnapshot interface {
+	// Name returns the name
+	Name() string
+
+	// Tags returns the tags
+	Tags() map[string]string
+
+	// Value returns the value
+	Value() int64
+}
+
+// GaugeSnapshot is a snapshot of a counter
+type GaugeSnapshot interface {
+	// Name returns the name
+	Name() string
+
+	// Tags returns the tags
+	Tags() map[string]string
+
+	// Value returns the value
+	Value() int64
+}
+
+// TimerSnapshot is a snapshot of a counter
+type TimerSnapshot interface {
+	// Name returns the name
+	Name() string
+
+	// Tags returns the tags
+	Tags() map[string]string
+
+	// Values returns the values
+	Values() []time.Duration
+}
+
+// StatsReporter is a backend for Scopes to report metrics to
+type StatsReporter interface {
+	// ReportCounter reports a counter value
+	ReportCounter(name string, tags map[string]string, value int64)
+
+	// ReportGauge reports a gauge value
+	ReportGauge(name string, tags map[string]string, value int64)
+
+	// ReportTimer reports a timer value
+	ReportTimer(name string, tags map[string]string, interval time.Duration)
+
+	// Capabilities returns a description of metrics reporting capabilities
+	Capabilities() Capabilities
+
+	// Flush is expected to be called by a Scope when it completes a round or reporting
+	Flush()
+}
+
+// Counter is the interface for logging statsd-counter-type metrics
+type Counter interface {
+	// Inc increments the counter by a delta
+	Inc(delta int64)
+}
+
+// Gauge is the interface for logging statsd-gauge-type metrics
+type Gauge interface {
+	// Update sets the gauges absolute value
+	Update(value int64)
+}
+
+// StopwatchStart is returned by a timer's start method, and should be passed
+// back to the timer's stop method at the end of the interval
+type StopwatchStart time.Time
+
+// Timer is the interface for logging statsd-timer-type metrics
+type Timer interface {
+	// Record Record a specific duration directly
+	Record(time.Duration)
+
+	// Start gives you back a specific point in time to report via Stop()
+	Start() StopwatchStart
```

Comment:
Once you return the stopwatch as an interface you cause the escape analysis of Go to allocate the stopwatch struct on the heap.  This causes allocation on the heap.

The only reason to avoid to returning just a plain exported struct is again the dependency management hell I described.